File: packages/sdk/src/economic/connectors/mpp.ts
/**
 * Stripe MPP (Machine Payments Protocol) connector.
 *
 * Reads MPP session spending limits and per-action costs from
 * session metadata. Tracks cumulative session spend against
 * Veto budget policies.
 *
 * Timing mode: pre-call — session metadata available at setup.
 * MPP spec version: March 2026 launch (spec may evolve).
 *
 * @module economic/connectors/mpp
 */

import type { EconomicContext, ProtocolConnector, EconomicDenialDetails } from '../types.js';

/**
 * MPP session data shape (from Stripe's session negotiation response).
 */
export interface MPPSessionData {
  /** Shared Payment Token identifier */
  session_token?: string;
  /** Per-action cost in the session currency */
  cost?: number;
  /** Session currency (ISO 4217) */
  currency?: string;
  /** Session spending limit */
  spending_limit?: number;
  /** Cumulative session spend so far */
  spent?: number;
  /** Payer identifier (Stripe customer ID or SPT holder) */
  payer?: string;
  /** Additional session metadata */
  [key: string]: unknown;
}

function extractMPPSessionData(obj: Record<string, unknown>): MPPSessionData | null {
  // Accept both top-level and nested mpp_session shapes
  const session = (obj.mpp_session ?? obj) as Record<string, unknown>;

  const sessionToken = session.session_token;
  if (typeof sessionToken !== 'string' || !sessionToken) return null;

  const cost = typeof session.cost === 'number' ? session.cost : undefined;
  const currency = typeof session.currency === 'string' ? session.currency : undefined;
  const spendingLimit = typeof session.spending_limit === 'number' ? session.spending_limit : undefined;
  const spent = typeof session.spent === 'number' ? session.spent : undefined;
  const payer = typeof session.payer === 'string' ? session.payer : undefined;

  return {
    session_token: sessionToken,
    cost,
    currency,
    spending_limit: spendingLimit,
    spent,
    payer,
  };
}

/**
 * Create an MPP protocol connector.
 */
export function createMPPConnector(): ProtocolConnector {
  return {
    protocol: 'mpp',
    protocolVersion: '2026.03',

    extract(response: Response | Record<string, unknown>): EconomicContext | null {
      // Handle fetch Response objects — MPP data comes in response headers/body
      if (response instanceof Response) {
        // Check for MPP headers
        const mppSession = response.headers.get('x-mpp-session');
        const mppCost = response.headers.get('x-mpp-cost');
        const mppCurrency = response.headers.get('x-mpp-currency');
        const mppPayer = response.headers.get('x-mpp-payer');

        if (!mppSession) return null;

        const cost = mppCost ? Number(mppCost) : undefined;
        if (cost !== undefined && (!Number.isFinite(cost) || cost < 0)) return null;

        return {
          cost: cost ?? 0,
          currency: (mppCurrency ?? 'USD').toUpperCase(),
          payer: mppPayer ?? undefined,
          protocol: 'mpp',
          protocol_metadata: {
            session_token: mppSession,
          },
        };
      }

      // Handle plain objects (manual mode / pre-call)
      const sessionData = extractMPPSessionData(response);
      if (!sessionData) return null;

      return {
        cost: sessionData.cost ?? 0,
        currency: (sessionData.currency ?? 'USD').toUpperCase(),
        payer: sessionData.payer,
        protocol: 'mpp',
        protocol_metadata: {
          session_token: sessionData.session_token,
          spending_limit: sessionData.spending_limit,
          session_spent: sessionData.spent,
        },
      };
    },
  };
}

/**
 * Build a connector_error denial for MPP parse failures.
 */
export function buildMPPConnectorError(error: string): EconomicDenialDetails {
  return {
    reason: 'connector_error',
    cost: 0,
    currency: 'USD',
    budget_scope: 'session',
    budget_limit: 0,
    budget_spent: 0,
    budget_remaining: 0,
    connector_name: 'mpp',
    raw_error: error,
  };
}

File: packages/sdk-python/veto/economic/connectors/mpp.py
"""Stripe MPP connector."""

from __future__ import annotations

import math
from collections.abc import Mapping
from typing import Any, Optional

from veto.economic.connectors._helpers import get_header
from veto.economic.types import EconomicContext, EconomicDenialDetails


def _number(value: Any) -> Optional[float]:
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        return None
    numeric = float(value)
    return numeric if math.isfinite(numeric) else None


def _extract_session_data(obj: Mapping[str, Any]) -> Optional[Mapping[str, Any]]:
    raw_session = obj.get("mpp_session", obj)
    if not isinstance(raw_session, Mapping):
        return None
    token = raw_session.get("session_token")
    if not isinstance(token, str) or not token:
        return None
    return raw_session


class MPPConnector:
    protocol = "mpp"
    protocol_version = "2026.03"
    protocolVersion = protocol_version

    def extract(self, response: Any) -> Optional[EconomicContext]:
        mpp_session = get_header(response, "x-mpp-session")
        if mpp_session:
            cost_header = get_header(response, "x-mpp-cost")
            cost = 0.0
            if cost_header is not None:
                try:
                    cost = float(cost_header)
                except ValueError:
                    return None
                if not math.isfinite(cost) or cost < 0:
                    return None

            currency = get_header(response, "x-mpp-currency") or "USD"
            payer = get_header(response, "x-mpp-payer")
            return EconomicContext(
                cost=cost,
                currency=currency.upper(),
                payer=payer,
                protocol="mpp",
                protocol_metadata={"session_token": mpp_session},
            )

        if not isinstance(response, Mapping):
            return None

        session = _extract_session_data(response)
        if session is None:
            return None

        session_cost = _number(session.get("cost"))
        session_currency = session.get("currency")
        session_payer = session.get("payer")
        return EconomicContext(
            cost=session_cost if session_cost is not None else 0,
            currency=session_currency.upper()
            if isinstance(session_currency, str)
            else "USD",
            payer=session_payer if isinstance(session_payer, str) else None,
            protocol="mpp",
            protocol_metadata={
                "session_token": session["session_token"],
                "spending_limit": _number(session.get("spending_limit")),
                "session_spent": _number(session.get("spent")),
            },
        )


def create_mpp_connector() -> MPPConnector:
    return MPPConnector()


def createMPPConnector() -> MPPConnector:
    return create_mpp_connector()


def build_mpp_connector_error(error: str) -> EconomicDenialDetails:
    return EconomicDenialDetails(
        reason="connector_error",
        cost=0,
        currency="USD",
        budget_scope="session",
        budget_limit=0,
        budget_spent=0,
        budget_remaining=0,
        connector_name="mpp",
        raw_error=error,
    )


def buildMPPConnectorError(error: str) -> EconomicDenialDetails:
    return build_mpp_connector_error(error)
